src/bindings/python/flux/constraint/parser.py

```diff
@@ -0,0 +1,453 @@
+        self.last_lparens = 0
+        self.last_rparens = 0
+        self.last_quote = None
+        self.quote_start = None
+        self.lexer.input(data)
+
+    def __getattr__(self, attr):
+        return getattr(self.lexer, attr)
+
+    def t_ANY_error(self, t):
+        raise ConstraintSyntaxError(
+            f"Illegal character '{t.value[0]}' at position {t.lexer.lexpos}"
+        )
+
+    #  Define special tokens as functions before t_TOKEN so they are
+    #  guaranteed to take precedence.
+    #  c.f. http://www.dabeaz.com/ply/ply.html#ply_nn6
+
+    def t_NEGATE(self, t):
+        r"-"
+        return t
+
+    def t_NOT(self, t):
+        r"not\b"
+        return t
+
+    def t_AND(self, t):
+        r"&{1,2}|and\b"
+        return t
+
+    def t_OR(self, t):
+        r"\|{1,2}|or\b"
+        return t
+
+    def t_LPAREN(self, t):
+        r"\("
+        self.parens_level += 1
+        self.last_lparens = t.lexer.lexpos - 1
+        return t
+
+    def t_RPAREN(self, t):
+        r"\)"
+        self.parens_level -= 1
+        self.last_rparens = t.lexer.lexpos - 1
+        return t
+
+    def t_TOKEN(self, t):
+        r"[^()|&\s\"\']+"
+        if self.pending_token is not None:
+            t.value = self.pending_token.value + t.value
+            self.pending_token = None
+        elif t.value.endswith(":"):
+            #  Save a token that ends with ':' to possibly combine with
+            #  any following token. This allows op:"quoted string"
+            self.pending_token = t
+            return None
+        return t
+
+    def t_eof(self, t):
+        if self.pending_token is not None:
+            val = self.pending_token.value
+            raise ConstraintSyntaxError(f"Missing argument to token '{val}'")
+        return None
+
+    def t_QUOTE(self, t):
+        r"'|\""  # fmt: skip
+        self.quote_start = t.lexer.lexpos
+        self.last_quote = t.lexer.lexpos - 1
+        if t.value == "'":
+            t.lexer.begin("squoting")
+        else:
+            t.lexer.begin("dquoting")
+
+    #  quoting state:
+    def t_squoting_TOKEN(self, t):
+        r"([^'])+"
+        return self.t_TOKEN(t)
+
+    def t_squoting_QUOTE(self, t):
+        r"'"
+        self.last_quote = None
+        t.lexer.begin("INITIAL")
+
+    def t_squoting_eof(self, t):
+        pos = self.quote_start
+        raise ConstraintSyntaxError(f'Unclosed quote "\'" at position {pos}')
+
+    def t_dquoting_TOKEN(self, t):
+        r"([^\"])+"
+        return self.t_TOKEN(t)
+
+    def t_dquoting_QUOTE(self, t):
+        r"\""  # fmt: skip
+        self.last_quote = None
+        t.lexer.begin("INITIAL")
+
+    def t_dquoting_eof(self, t):
+        pos = self.quote_start
+        raise ConstraintSyntaxError(f"Unclosed quote '\"' at position {pos}")
+
+
+class ConstraintParser:
+    r"""
+    Base constraint query syntax parser class.
+
+    This class implements the simple constraint query language with
+    the following grammar:
```

Comment:
Should this be introduced as RFC 35 constraint syntax in the comment block so future hackers know that the implementation is constrained by that document?

Also, not sure if "value" should be referred to as "operand" here?

Reply:
Good point. I forgot to go back and reword the class description after RFC 35 was finalized.

Reply:
Description updated.